### PR TITLE
ShaderLab support MRT

### PR DESCRIPTION
**Pull request by Sway007**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

This pr fullfilled issue #2373 

## Summary by CodeRabbit

- **New Features**
	- Introduced multi-render target (MRT) shader implementation.
	- Added new configuration property for MRT in the E2E configuration.
	- Enhanced grammar to support layout qualifiers in shader programs.
	- Added built-in variable `gl_FragData` for fragment shaders.
	- Implemented new methods for symbol management in the symbol table.

- **Bug Fixes**
	- Improved error handling and reporting mechanisms in various components.

- **Tests**
	- Added new test cases for MRT validation and error checking.

- **Documentation**
	- Updated class documentation and comments for clarity.

- **Refactor**
	- Simplified class constructors and method signatures for better maintainability.
	- Restructured error reporting logic across multiple classes.
	- Updated type references from `ENonTerminal` to `NoneTerminal` across the parser components.

### CODE REVIEW

File: packages/shader-lab/src/codeGen/GLES300.ts [hunk 1]
```diff
@@ -129,4 +126,27 @@ export class GLES300Visitor extends GLESVisitor {
...
+      if (!_referencedVaryingList[V3_GL_FragColor]) {
+        const token = Token.pool.get();
+        token.set(ETokenType.ID, V3_GL_FragColor, ShaderLab.createPosition(0, 0, 0));
+        _referencedVaryingList[V3_GL_FragColor] = {
+          ident: token,
+          typeInfo: new SymbolType(EKeyword.VEC4, "vec4"),
+          qualifier: "out",
+          astNode: node
+        };
+      }
```
Comment (zhuxudong) [inline packages/shader-lab/src/codeGen/GLES300.ts:145]: Encapsulate this logic that repeats `visitVariableIdentifier` into a function, such as `registerFragColorVariable`

Comment (Sway007) [reply packages/shader-lab/src/codeGen/GLES300.ts]: done

File: packages/shader-lab/src/parser/AST.ts [hunk 2]
```diff
@@ -13,7 +13,23 @@ import { IParamInfo, NodeChild, StructProp, SymbolType } from "./types";
...
 
+function ASTNodeDecorator(nonterminal: ENonTerminal) {
+  return function <T extends { new (...args: any[]): TreeNode }>(ASTNode: T) {
+    const ASTNodeClass = class extends ASTNode {
+      override set(loc: ShaderRange, children: NodeChild[]): void {
+        super.set(loc, children, nonterminal);
+        super.init();
+      }
+    };
+    // @ts-ignore
```
Comment (GuoLei1990) [inline packages/shader-lab/src/parser/AST.ts:24]: Why add ingnore

File: packages/shader-lab/src/GSError.ts [hunk 3]
```diff
@@ -47,7 +47,8 @@ export class GSError extends Error {
       if (i === start.line) {
         remarkStart = start.column;
         paddingLength += start.column;
-      } else if (i === end.line) {
+      }
+      if (i === end.line) {
```
Comment (GuoLei1990) [inline packages/shader-lab/src/GSError.ts:51]: Maybe start and end  at the same time?

**Pull request merged**